Python3 cleanup

dannon opened the pull request:
Now that we're officially py3-only, we can clean up the cruft and standardize a few things.

I used some upgrade tooling to do stuff here like swap to a single class style, drop `six` compatibility shims in some places where it was straightforward, standardize print formatting and drop __future__ imports, etc.  There's a good bit more to do, some tweaky cases that we'll want to handle manually, but I wanted to draft this and get tests running.

dannon committed:
Python3 upgrade pass, part 1; basics.

dannon committed:
Swap percent formattings to modern .format

dannon committed:
More python 3 upgrade; this time python3+ feature rewrite, drop six, etc

dannon committed:
Missed one new style class here.

dannon committed:
Update rest of lib with dict comprehensions, etc.

dannon committed:
Update string formatting in rest of lib

dannon committed:
Remainder of py3-only feature updates.

dannon commented on lib/galaxy/datatypes/converters/bed_to_gff_converter.py:
```diff
@@ -1,6 +1,5 @@
 #!/usr/bin/env python
 # This code exists in 2 places: ~/datatypes/converters and ~/tools/filters
-from __future__ import print_function
```
I might need to back out tweaks like this, and other converters, that can/should still run with an older python?

nsoranzo replied:
The tool XML already has `<requirement type="package" version="3.7">python</requirement>`, so I think it's fine.

dannon committed:
Drop empty 'pass', cleanup some unused imports

dannon committed:
Drop now-unused six imports

dannon committed:
Fix continuation line over indented

dannon committed:
A pass through tool_shed linting

dannon committed:
Fix empty lines

dannon committed:
galaxy.tools.parameters remaining fixes; swap to io.stringio

dannon committed:
Swap galaxy.datatypes.sniff to urllib and io usage

dannon committed:
Fix remaining tool_util linting/conversion issues.

dannon committed:
Fix remaining tool_util linting/conversion issues, minor tweak to
tool_shed file handling.

dannon committed:
Import order, forgot alphabet.

dannon committed:
Drop PY2 specific doctest

dannon committed:
Fix dynamic tool destination handling of infinity exception

dannon committed:
Drop py2 exception and update http.cookies in base webapp

dannon committed:
Merge remote-tracking branch 'upstream/dev' into py3-compat-drop

dannon committed:
Post-merge pass updating py3 features/standardizing

dannon committed:
Drop unused import; swap text_type to str for py3 (smart_str/etc
probably needs a full pass now for being py3 only)

dannon committed:
Test fix next iteration for py3 without six

dannon committed:
Fix another dict view/values iteration

dannon committed:
Merge remote-tracking branch 'upstream/dev' into py3-compat-drop

dannon committed:
Merge remote-tracking branch 'upstream/dev' into py3-compat-drop

dannon committed:
Swap tool_util to py3 standard string formatting

nsoranzo commented on lib/galaxy/datatypes/converters/maf_to_fasta_converter.py:
```diff
@@ -16,7 +15,7 @@ def __main__():
     input_name = sys.argv.pop(1)
     count = 0
     with open(output_name, 'w') as out:
```
```suggestion
    with open(output_name, 'w') as out, open(input_name) as in:
```

nsoranzo commented on lib/galaxy/datatypes/converters/maf_to_fasta_converter.py:
```diff
@@ -16,7 +15,7 @@ def __main__():
     count = 0
     with open(output_name, 'w') as out:
-        for count, block in enumerate(bx.align.maf.Reader(open(input_name, 'r'))):
+        for count, block in enumerate(bx.align.maf.Reader(open(input_name))):
```
```suggestion
        for count, block in enumerate(bx.align.maf.Reader(in)):
```

nsoranzo commented on lib/galaxy/datatypes/data.py:
```diff
@@ -488,7 +483,7 @@ def display_as_markdown(self, dataset_instance, markdown_format_helpers):
             result = "*cannot display binary content*\n"
         else:
-            contents = open(dataset_instance.file_name, "r").read(DEFAULT_MAX_PEEK_SIZE)
+            contents = open(dataset_instance.file_name).read(DEFAULT_MAX_PEEK_SIZE)
```
```suggestion
            with open(dataset_instance.file_name) as f:
                contents = f.read(DEFAULT_MAX_PEEK_SIZE)
```

nsoranzo commented on lib/galaxy/datatypes/data.py:
```diff
@@ -506,7 +501,7 @@ def _yield_user_file_content(self, trans, from_dataset, filename):
             # TODO Ideally this happens a layer higher, but this is a bad
             # issue affecting many tools
-            return sanitize_html(open(filename, 'r').read()).encode('utf-8')
+            return sanitize_html(open(filename).read()).encode('utf-8')
```
```suggestion
            with open(filename) as f:
                return sanitize_html(f.read()).encode('utf-8')
```

nsoranzo commented on lib/galaxy/datatypes/display_applications/parameters.py:
```diff
@@ -190,7 +190,7 @@ def url(self):
             base_url = "http%s" % base_url[5:]
-        return "%s%s" % (base_url,
+        return "{}{}".format(base_url,
                          self.trans.app.url_for(controller='dataset',
```
Indentation seems wrong here.

nsoranzo commented on lib/galaxy/datatypes/genetics.py:
```diff
@@ -712,7 +712,7 @@ def set_peek(self, dataset, **kwd):
             pp = os.path.join(dataset.extra_files_path, '%s.pheno' % dataset.metadata.base_name)
             try:
-                p = open(pp, 'r').readlines()
+                p = open(pp).readlines()
```
```suggestion
                with open(pp) as f:
                    p = f.readlines()
```

nsoranzo commented on lib/galaxy/datatypes/genetics.py:
```diff
@@ -727,7 +727,7 @@ def get_peek(self, dataset):
         pp = os.path.join(dataset.extra_files_path, '%s.pheno' % dataset.metadata.base_name)
         try:
-            p = open(pp, 'r').readlines()
+            p = open(pp).readlines()
```
```suggestion
            with open(pp) as f:
                p = f.readlines()
```

nsoranzo commented on lib/galaxy/datatypes/genetics.py:
```diff
@@ -738,7 +738,7 @@ def get_file_peek(self, filename):
         h = '## rexpression get_file_peek: no file found'
         try:
-            h = open(filename, 'r').readlines()
+            h = open(filename).readlines()
```
```suggestion
            with open(filename) as f:
                h = f.readlines()
```

nsoranzo commented on lib/galaxy/datatypes/genetics.py:
```diff
@@ -788,7 +788,7 @@ def set_meta(self, dataset, **kwd):
         dataset.metadata.pheno_path = pp
         try:
-            pf = open(pp, 'r').readlines()  # read the basename.phenodata in the extra_files_path
+            pf = open(pp).readlines()  # read the basename.phenodata in the extra_files_path
```
```suggestion
            with open(pp) as f:
                pf = f.readlines()  # read the basename.phenodata in the extra_files_path
```

nsoranzo commented on lib/galaxy/datatypes/sequence.py:
```diff
@@ -780,7 +778,7 @@ def process_split_file(data):
         sequence_count = long(args['num_sequences'])
```
```suggestion
        sequence_count = int(args['num_sequences'])
```

nsoranzo commented on lib/galaxy/datatypes/sequence.py:
```diff
@@ -36,8 +35,7 @@
 
-if sys.version_info > (3,):
-    long = int
+long = int
```
Should we replace `long` with `int` in the file instead?

dannon replied:
I went ahead and did this across the codebase; good idea.

nsoranzo commented on lib/galaxy/datatypes/sequence.py:
```diff
@@ -780,7 +778,7 @@ def process_split_file(data):
 
         if 'toc_file' in args:
-            toc_file = json.load(open(args['toc_file'], 'r'))
+            toc_file = json.load(open(args['toc_file']))
```
```suggestion
            with open(args['toc_file']) as f:
                toc_file = json.load(f)
```

nsoranzo commented on lib/galaxy/datatypes/sniff.py:
```diff
@@ -13,16 +12,9 @@
 import shutil
 import sys
 import tempfile
+import urllib
```
```suggestion
import urllib.request
```

nsoranzo commented on lib/galaxy/datatypes/tabular.py:
```diff
@@ -30,8 +28,7 @@
 
-if sys.version_info > (3,):
-    long = int
+long = int
```
Should we replace `long` with `int` in the file instead?

nsoranzo commented on lib/galaxy/datatypes/tabular.py:
```diff
@@ -125,7 +122,7 @@ def display_data(self, trans, dataset, preview=False, filename=None, to_ext=None
 
     def display_as_markdown(self, dataset_instance, markdown_format_helpers):
-        contents = open(dataset_instance.file_name, "r").read(data.DEFAULT_MAX_PEEK_SIZE)
+        contents = open(dataset_instance.file_name).read(data.DEFAULT_MAX_PEEK_SIZE)
```
```suggestion
        with open(dataset_instance.file_name) as f:
            contents = f.read(data.DEFAULT_MAX_PEEK_SIZE)
```

nsoranzo commented on lib/galaxy/datatypes/tabular.py:
```diff
@@ -1002,7 +998,7 @@ def sniff(self, filename):
         """ Return True if if recognizes dialect and header. """
         try:
```
We can get rid of `try:` and the `except` block.

nsoranzo commented on lib/galaxy/datatypes/tabular.py:
```diff
@@ -1002,7 +998,7 @@ def sniff(self, filename):
         try:
             # check the dialect works
-            reader = csv.reader(open(filename, 'r'), self.dialect)
+            reader = csv.reader(open(filename), self.dialect)
```
```suggestion
            with open(filename) as f:
                reader = csv.reader(f, self.dialect)
```

nsoranzo commented on lib/galaxy/datatypes/tabular.py:
```diff
@@ -1031,7 +1027,7 @@ def sniff(self, filename):
 
             # Optional: Check Python's csv comes up with a similar dialect
-            auto_dialect = csv.Sniffer().sniff(open(filename, 'r').read(self.big_peek_size))
+            auto_dialect = csv.Sniffer().sniff(open(filename).read(self.big_peek_size))
```
```suggestion
            with open(filename) as f:
                big_peek = f.read(self.big_peek_size)
            auto_dialect = csv.Sniffer().sniff(big_peek)
```

nsoranzo commented on lib/galaxy/datatypes/tabular.py:
```diff
@@ -1046,7 +1042,7 @@ def sniff(self, filename):
                   this test may be checking the wrong dialect.
             """
-            if not csv.Sniffer().has_header(open(filename, 'r').read(self.big_peek_size)):
+            if not csv.Sniffer().has_header(open(filename).read(self.big_peek_size)):
```
```suggestion
            if not csv.Sniffer().has_header(big_peek):
```

nsoranzo reviewed (commented):
Reviewed around a 10%, I will be on holiday from tomorrow so submitting what I have.

dannon committed:
Apply suggestions from code review

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

dannon committed:
Fix indentation in display_applications

dannon committed:
Merge remote-tracking branch 'upstream/dev' into py3-compat-drop

dannon committed:
Swap long() for int() for python3 shift.

nsoranzo committed:
Don't use reserved words as variable name

nsoranzo committed:
File needs to stay open while read via `csv.reader()`

Also:
- drop unnecessary try/except
- open CSV files with `newline=''`, as recommended in https://docs.python.org/3/library/csv.html#csv.reader

bgruening commented on lib/galaxy/managers/citations.py:
```diff
@@ -146,10 +146,10 @@ def to_bibtex(self):
-                DOI = {%s},
-                note = {Failed to fetch BibTeX for DOI.}
-            }""" % (self.__doi, self.__doi)
+            return """@MISC{{{},
```
This is not more readable than before. Its worse, hopefully we can soon use f-strings.

dannon replied:
Yeah, this particular one is maybe worse with all the other braces involved.   Want me to update it to have the variable names inside?  That might help here.  (done in https://github.com/galaxyproject/galaxy/pull/9923/commits/a2d1ec58040c9ee2eb296ab7d03f8c8415c40a29)

I actually *did* do f-string conversion at one point, so the good news is that it's very easy to do when we're ready.  The problem is that it's 3.6+ and I think we support 3.5.

bgruening replied:
Jupp, we need at first to drop 3.5. So we should wait. We need to fix uwsgi-support for 3.7, 3.8+ so we can also use those python version :(

> Want me to update it to have the variable names inside?

I think that would be nice.

bgruening commented on lib/galaxy/managers/history_contents.py:
```diff
@@ -397,7 +397,7 @@ def _contained_id_map(self, id_list):
             .options(eagerload('tags'))
             .options(eagerload('annotations')))
-        return dict((row.id, row) for row in query.all())
+        return {row.id: row for row in query.all()}
```
I would prefer dict() and it is supposed to be faster than {} ... but I guess we do not care much?

dannon replied:
Weird, I saw the comprehension format as faster in a quick test.  What situations is it not supposed to be good for?

```
[yoplait@incus:~/work/galaxy] python3 -m timeit -s 'import string' 'b = {i: 1 for i in string.digits}'
500000 loops, best of 5: 617 nsec per loop
[yoplait@incus:~/work/galaxy] python3 -m timeit -s 'import string' 'b = dict((i,1) for i in string.digits)'
200000 loops, best of 5: 1.33 usec per loop
```

bgruening replied:
Mh, ok, interesting. Lets leave it like that. Thanks.

bgruening reviewed (commented):
review included up until  lib/galaxy/model/item_attrs.py

dannon committed:
Make citation manager string formatting more readable

dannon committed:
Merge remote-tracking branch 'upstream/dev' into py3-compat-drop

dannon committed:
Post-merge fixups

dannon committed:
Merge remote-tracking branch 'upstream/dev' into py3-compat-drop

bgruening commented on lib/galaxy/util/object_wrapper.py:
```diff
@@ -108,10 +106,11 @@
-    def coerce(x, y):
-        return x
+# __coerce__ doesn't do anything under Python anyway.
+
```
empty lines could be removed or removed completely?

dannon replied:
Yeah, I put the comment inside the function, thank you!

bgruening commented on lib/galaxy/util/rules_dsl.py:
```diff
@@ -73,7 +70,7 @@ class AddColumnMetadataRuleDefinition(BaseRuleDefinition):
 
     def validate_rule(self, rule):
-        _ensure_rule_contains_keys(rule, {"value": six.string_types})
+        _ensure_rule_contains_keys(rule, {"value": (str,)})
```
can this be `str`?

dannon replied:
Yep, sure can.  I'll swap it.

bgruening reviewed (approved):
Ok, I looked at all files starting with @nsoranzo ones. Just two small questions but please feel free to merge. Those rebeases are probably annoying.

And we can remove six from all our requirements files.

dannon committed:
Use simple string instead of tuple in rules_dsl (conversion from six.string_types)

dannon committed:
Put comment in function instead of two lines above

dannon committed:
Merge remote-tracking branch 'upstream/dev' into py3-compat-drop
